### ISSUE TITLE
`devDependencies` option

### ISSUE THREAD

**Issue by starper** [spacchetti/spago#80]
What do you think about adding `devDependencies` option (to both `Package` and `Config`)? The motivation here is that most test frameworks are pretty big, with lots of transitive dependencies and different libs can use different test frameworks, wich adds even more packages that I will have to download, but most of the time will not use. Downloading only those packages that are required for library to work should make installation much faster.

**Comment by f-f** [comment]
I considered this at some point, and I came to the conclusion it's not a feature I'd like to have.
Some clarifications on the why and how of all of this:
- first, I'll quickly note that the `Package` type is not in our hands, but comes from the [spacchetti](https://github.com/spacchetti/spacchetti) package-set. The list of `Package`s is only the list of *available* packages that you can install, so they are not really affecting anything in this case
- about the `Config`: in general I've always been confused by how `npm` has `dependencies` and `devDependencies`, they cause much more trouble than they save, as I always mix them up
- in general, if the download is unbearably slow we should focus on making download faster by some order of magnitude, instead of trying to cut the download time in two. This would also complicate the logic of downloading things (because now you have to places that tell you what to download), and I don't really feel like it, since it's complicated enough to maintain
- the "big" download should happen just once per project, when you install it for the first time. There's virtually no reason to remove the `.spago` folder at all, since we download stuff there by tag (which are effectively immutable refs)

**Comment by starper** [comment]
> I came to the conclusion it's not a feature I'd like to have.

Ok. I guess you can close this issue then.

**Comment by f-f** [comment]
@starper though I'd be interested if you have such an example of a configuration that takes a lot of time to install

**Comment by f-f** [comment]
Cool, then we'll have this issue to reference if anyone hits that use case. Thank you! 🙂

**Comment by milesfrain** [comment]
It looks like a good workaround has been posted to the [devDependencies section](https://github.com/purescript/spago#devdependencies-testdependencies-or-in-general-a-situation-with-many-configurations) of the readme. Figured I'd post a reference to it here, since this issue is the first to appear when searching for "spago dev dependencies".

**Comment by Hi-Angel** [comment]
@milesfrain the link is broken

@f-f related to [this question](https://discourse.purescript.org/t/spago-0-20-released/2283/3?u=hi-angel), how do you see being able to suppress the "unused deps" Spago warning without a `devDependencies` option?

**Comment by Hi-Angel** [comment]
Ideas on the workarounds are welcome. I tried to work around that by creating a special `DevDeps.purs` file, where the deps would be "used", but that results then in warnings "Debug function usage".

**Comment by f-f** [comment]
@Hi-Angel we have a `test.dependencies` key in the config, which will install the dependencies when you build locally, while not including them for downstream packages that depend on your library.

**Comment by Hi-Angel** [comment]
> @Hi-Angel we have a `test.dependencies` key in the config, which will install the dependencies when you build locally, while not including them for downstream packages that depend on your library.

Unless I'm doing something wrong, it doesn't seem to work.

I added this line to `spago.dhall`:

```json
, test.dependencies = [ "debug", "undefined" ]
```

…then I removed `"debug"` from the main `dependencies = ` key and added a `spy` usage in the sources.

Executing `spago build` afterwards result in 

```haskell
[error] Some of your project files import modules from packages that are not in the direct dependencies of your project.
To fix this error add the following packages to the list of dependencies in your config:
- debug
You may add these dependencies by running the following command:
spago install debug
```

**Comment by Hi-Angel** [comment]
> You are using the [legacy](https://github.com/purescript/spago-legacy) version of spago, which is lacking many of the features of `spago-next`, which is being developed in this repo.

Actually no, I'm using the purescript-written Spago, latest released 0.21.0. I saw the older was deprecated when I was starting with purescript, so I installed [it from AUR](https://aur.archlinux.org/packages/nodejs-spago), and [releases page doesn't have anything after 0.21.0](https://github.com/purescript/spago/releases), so there's nothing to update to.

I think you meant to say that `test.dependencies` feature is on the master branch and will be available in the next release. Okay.

**Comment by Hi-Angel** [comment]
Fun fact btw, [spago-releases has more releases](https://github.com/purescript/spago-legacy/releases), than the rewrite.

**Comment by f-f** [comment]
> Actually no, I'm using the purescript-written Spago, latest released 0.21.0

As the README explains, the releases up to the 0.21.x line are from the old Haskell codebase.
The purescript codebase covers the releases from the 0.90.x line. They are released only on NPM as of now, and you can get it with `npm install -g spago@next`.

**Comment by f-f** [comment]
> Fun fact btw, [spago-legacy has more releases](https://github.com/purescript/spago-legacy/releases), than the rewrite.

Yes, I cut an additional release after a request from the nixpkgs maintainers.
